### PR TITLE
Fix newline rejection and whitespace trimming for chat messages 

### LINKED ISSUE #13916: `register_on_chat_message` gets a whitespace-trimmed chat message
### Minetest version

```true
Minetest 5.8.0-dev-6fdc7e0da-dirty (Linux)
```

### Active renderer

_No response_

### Irrlicht device

_No response_

### Operating system and version

arch linux

### CPU model

_No response_

### GPU model

_No response_

### OpenGL version

_No response_

### Summary

See `Server::handleChat`:
https://github.com/minetest/minetest/blob/b0107144261f6513bbde95323c9663a71ed8f3c9/src/server.cpp#L3035

`core.format_chat_message` receives a message with whitespace, `core.format_chat_message` with trimmed whitespace.

Apparently #8420 added the trimming. cc @rubenwardy 
Is there some rationale behind this (not sure if it's a bug)?

### Steps to reproduce

```lua
core.register_on_chat_message(function(name, message)
	core.log("on chat msg: "..dump(message))
end)
```

In `builtin/game/chat.lua`:
```diff
function core.format_chat_message(name, message)
+	core.log("msg to format: "..dump(message))
```

Write e.g. ` foo `.
=> `"foo"` vs. " foo "

### Additional context

I found this because if you enable csm, you get the "Issued chat command: ..." messages, but not if you put whitespace before the `/`.

#### Offtopic

I had to create this Additional context section inside the Steps to reproduce field. I hate the issue form, especially the current one.

### PR DESCRIPTION
**Pull request by chmodsayshello**
Prior to this commit, the trimmed string was checked for the forbidden characters, which could lead to false negatives.

closes #13916

# Goal of the PR
- Enforce the use of allowed characters

## How does the PR work?
- Check the original wstring instead

**Ready for Review.**

## How to test

**note**: outdated

- Load up both the latest master and this branch of minetest
- start singleplayer games or HOST a server in both versions
- send the following chat messages (replace \r and \x1b with the corresponding characters):

>  `\r\r\r\r\r\r\r\r\r\r\rHello, this message has multiple lines`
 `\x1b(c@#ff0000)This is a red chat message.`

- Verify that the current master happily accepts all of these and my branch does not.

## Screenshots

| Currently | With this PR |
|--------------|------------------|
| grafik  | grafik |
| grafik |  grafik |
| Yes, this is actually a single chat message, the formatting code can just screw up badly with carriage returns. grafik |  grafik |

### CODE REVIEW

File: src/server.h [hunk 1]
```diff
@@ -50,6 +50,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <shared_mutex>
 #include <condition_variable>
 
+// U+001B is the Escape character needed for color codes
+#define CHAT_DISALLOWED_CHARS L"\n\r\x1b"
```
Comment (sfan5) [inline src/server.h:54]: but isn't this covered by `strip_color_codes`?

File: src/server.cpp [hunk 2]
```diff
@@ -3091,8 +3091,8 @@ std::wstring Server::handleChat(const std::string &name,
 	if (message.empty())
 		return L"";
 
-	if (message.find_first_of("\n\r") != std::wstring::npos) {
-		return L"Newlines are not permitted in chat messages";
+	if (wmessage.find_first_of(CHAT_DISALLOWED_CHARS) != std::wstring::npos) {
```
Comment (sfan5) [inline src/server.cpp:3094]: shouldn't we be removing the trim instead?

**Comment by sfan5** [comment]
I've updated the PR according to my own review. It should still work.

File: src/server.cpp [hunk 3]
```diff
@@ -3087,13 +3089,12 @@ std::wstring Server::handleChat(const std::string &name,
...
-	auto message = trim(wide_to_utf8(wmessage));
+	auto message = wide_to_utf8(wmessage);
 	if (message.empty())
 		return L"";
 
-	if (message.find_first_of("\n\r") != std::wstring::npos) {
-		return L"Newlines are not permitted in chat messages";
-	}
+	if (message.find_first_of("\r\n") != std::string::npos)
+		return L"Your message contains disallowed characters.";
```
Comment (grorp) [inline src/server.cpp:3097]: old error message is more helpful

**Comment by grorp** [comment]
This would be a nice thing to have a simple unit test for.

File: src/server.cpp [hunk 4]
```diff
@@ -3115,8 +3116,7 @@ std::wstring Server::handleChat(const std::string &name,
 #ifdef __ANDROID__
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/server.cpp:3116]: I agree, but out-of scope

File: src/server.cpp (hunk 3 above)
Comment (sfan5) [inline src/server.cpp:3097]:
```suggestion
		return L"Newlines are not permitted in chat messages";
```

**Comment by sfan5** [comment]
> This would be a nice thing to have a simple unit test for.

I don't see how.

**Review by Ekdohibs — APPROVED** [review]

**Review by grorp — APPROVED** [review]
`\r\r\r\r\rHello, this message\r\r\r\r\r\rhas multiple lines` is rejected

`\r\r\r\r\rHello, this message has multiple lines` is not rejected, but the leading whitespace is trimmed

Works 👍

**Pull request merged**